Use `Writable` utility type to improve typings

JeanMeche opened the pull request:
The `Writable` type is usefull when we want overwrite readonly properties and we still want to maintain code navigation/reference. It should be use instead of `any` or `{myPrivateKey: any}` type assertions for example.

See individual commits

## PR Type
What kind of change does this PR introduce?

- [x] Refactoring (no functional changes, no api changes)


## Does this PR introduce a breaking change?

- [x] No

JeanMeche committed:
refactor(core): Private export of the `Writable` type

The `Writable` type is usefull when we want overwrite readonly properties and we still want to maintain code navigation/reference. It should be use instead of `any` type assertion for example.

JeanMeche committed:
refactor(forms): Use the `Writable` type when overwriting readonly properties.

The `Writable` type is usefull when we want overwrite readonly properties and we still want to maintain code navigation/reference. It should be use instead of `any` type assertions for example.

JeanMeche committed:
refactor(core): Use the `Writable` type when overwriting readonly properties.

The `Writable` type is usefull when we want overwrite readonly properties and we still want to maintain code navigation/reference. It should be use instead of `any` type assertions for example.

JeanMeche committed:
refactor(router): Use the `Writable` type when overwriting readonly properties.

The `Writable` type is usefull when we want overwrite readonly properties and we still want to maintain code navigation/reference. It should be use instead of `any` type assertions for example.

JeanMeche committed:
refactor(platform-server): Use the `Writable` type when overwriting readonly properties.

The `Writable` type is usefull when we want overwrite readonly properties and we still want to maintain code navigation/reference. It should be use instead of `any` type assertions for example.

JeanMeche commented on packages/forms/src/directives/ng_control_status.ts:
```diff
@@ -54,7 +57,7 @@ export class AbstractControlStatus {
     // We check for the `submitted` field from `NgForm` and `FormGroupDirective` classes, but
     // we avoid instanceof checks to prevent non-tree-shakable references to those types.
-    return !!(this._cd as unknown as {submitted: boolean} | null)?.submitted;
+    return !!(this._cd as Writable<NgForm|FormGroupDirective>| null)?.submitted;
```
This should break the treeshaking right ?

garrettld replied:
You could use type-only imports to be certain that it won't, e.g. `import { type NgForm } from './ng_form'`.

JeanMeche replied:
Great idea, thanks !

JeanMeche commented on packages/core/src/linker/query_list.ts:
```diff
@@ -154,16 +155,13 @@ export class QueryList<T> implements Iterable<T> {
-    // QueryList (but not for QueryList itself.)
-    const self = this as QueryListInternal<T>;
-    (self as {dirty: boolean}).dirty = false;
+    (this as {dirty: boolean}).dirty = false;
```
Any idea why this property is of type `true` and not a boolean ?

JeanMeche committed:
refactor(animations): Use the `Writable` type when overwriting readonly properties.

The `Writable` type is usefull when we want overwrite readonly properties and we still want to maintain code navigation/reference. It should be use instead of  type assertions.

dylhunn commented:
This is a nice change conceptually. I just need to sit down and look over all the usage points.

JeanMeche commented:
@dylhunn Should be really straight forward 😊

JeanMeche commented:
The presubmit is failing, any info on what is breaking ?

jessicajaniuk reviewed (approved):
reviewed-for: fw-animations

dylhunn reviewed (approved):
reviewed-for: fw-core, fw-forms

dylhunn commented:
g3 presubmit is passing.

dylhunn commented:
disabling pullapprove -- router change is a no-op

dylhunn commented:
This PR was merged into the repository by commit e5d327d0496d23578cf09b548f7c3a14774407fc.
